# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/resistfingerprinting/tests/test_reduceprecision.cpp
# Commit: 85317d582166
# Full Hash: 85317d5821665149ffa385e8b55b244565c9d4c4
# Author: Tom Tung <ttung@mozilla.com>
# Date: 2020-04-07 21:44:02
# Regressor Bug: 1586761
# File Overlap Count: 1
# Description:
#   Bug 1586761 - P2 - Introduce new TimerPrecisionTypes and a set of new Reduce methods to decide the TimerPrecisionType in the nsRFPService; r=tjr
#   
#   To support checking CrossOriginIsolated in performance.now(), we need to:
#   - Add new types to TimerPrecisionType for nsRFPService
#     - System, HighResAllowed are added
# ==============================================================================

diff -r b51663641450 -r 85317d582166 toolkit/components/resistfingerprinting/tests/test_reduceprecision.cpp
--- a/toolkit/components/resistfingerprinting/tests/test_reduceprecision.cpp	Wed Mar 25 13:53:46 2020 +0000
+++ b/toolkit/components/resistfingerprinting/tests/test_reduceprecision.cpp	Wed Mar 25 14:09:55 2020 +0000
@@ -73,9 +73,9 @@
 void process(double clock, nsRFPService::TimeScale clockUnits,
              double precision) {
   double reduced1 = nsRFPService::ReduceTimePrecisionImpl(
-      clock, clockUnits, precision, -1, TimerPrecisionType::All);
+      clock, clockUnits, precision, -1, TimerPrecisionType::Normal);
   double reduced2 = nsRFPService::ReduceTimePrecisionImpl(
-      reduced1, clockUnits, precision, -1, TimerPrecisionType::All);
+      reduced1, clockUnits, precision, -1, TimerPrecisionType::Normal);
   ASSERT_EQ(reduced1, reduced2);
 }
 
@@ -133,31 +133,124 @@
   double result;
   result = nsRFPService::ReduceTimePrecisionImpl(
       2611.14, nsRFPService::TimeScale::MilliSeconds, 20, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 2611.14);
   result = nsRFPService::ReduceTimePrecisionImpl(
       2611.145, nsRFPService::TimeScale::MilliSeconds, 20, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.141, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::Normal);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.15999, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::Normal);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.15, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::Normal);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.13, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::Normal);
+  ASSERT_EQ(result, 2611.12);
+  cleanupJitter(jitterEnabled);
+}
+
+TEST(ResistFingerprinting, ReducePrecision_Expectations_HighRes)
+{
+  bool jitterEnabled = setupJitter(false);
+  double result;
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.14, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.145, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
   ASSERT_EQ(result, 2611.14);
   result = nsRFPService::ReduceTimePrecisionImpl(
       2611.141, nsRFPService::TimeScale::MilliSeconds, 20, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.15999, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.15, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.13, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 2611.12);
+  cleanupJitter(jitterEnabled);
+}
+
+TEST(ResistFingerprinting, ReducePrecision_Expectations_RFP)
+{
+  bool jitterEnabled = setupJitter(false);
+  double result;
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.14, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::RFP);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.145, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::RFP);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.141, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::RFP);
   ASSERT_EQ(result, 2611.14);
   result = nsRFPService::ReduceTimePrecisionImpl(
       2611.15999, nsRFPService::TimeScale::MilliSeconds, 20, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::RFP);
   ASSERT_EQ(result, 2611.14);
   result = nsRFPService::ReduceTimePrecisionImpl(
       2611.15, nsRFPService::TimeScale::MilliSeconds, 20, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::RFP);
   ASSERT_EQ(result, 2611.14);
   result = nsRFPService::ReduceTimePrecisionImpl(
       2611.13, nsRFPService::TimeScale::MilliSeconds, 20, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::RFP);
   ASSERT_EQ(result, 2611.12);
   cleanupJitter(jitterEnabled);
 }
 
+TEST(ResistFingerprinting, ReducePrecision_Expectations_DangerouslyNone)
+{
+  bool jitterEnabled = setupJitter(false);
+  double result;
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.14, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 2611.14);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.145, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 2611.145);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.141, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 2611.141);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.15999, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 2611.15999);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.15, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 2611.15);
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      2611.13, nsRFPService::TimeScale::MilliSeconds, 20, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 2611.13);
+  cleanupJitter(jitterEnabled);
+}
+
 TEST(ResistFingerprinting, ReducePrecision_ExpectedLossOfPrecision)
 {
   bool jitterEnabled = setupJitter(false);
@@ -165,27 +258,112 @@
   // We lose integer precision at 9007199254740992 - let's confirm that.
   result = nsRFPService::ReduceTimePrecisionImpl(
       9007199254740992.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 9007199254740990.0);
   // 9007199254740995 is approximated to 9007199254740996
   result = nsRFPService::ReduceTimePrecisionImpl(
       9007199254740995.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 9007199254740996);
   // 9007199254740999 is approximated as 9007199254741000
   result = nsRFPService::ReduceTimePrecisionImpl(
       9007199254740999.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 9007199254741000.0);
   // 9007199254743568 can be represented exactly, but will be clamped to
   // 9007199254743564
   result = nsRFPService::ReduceTimePrecisionImpl(
       9007199254743568.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 9007199254743564.0);
   cleanupJitter(jitterEnabled);
 }
 
+TEST(ResistFingerprinting, ReducePrecision_ExpectedLossOfPrecision_HighRes)
+{
+  bool jitterEnabled = setupJitter(false);
+  double result;
+  // We lose integer precision at 9007199254740992 - let's confirm that.
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740992.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 9007199254740980.0);
+  // 9007199254740995 is approximated to 9007199254740980
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740995.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 9007199254740980);
+  // 9007199254740999 is approximated as 9007199254741000
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740999.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 9007199254741000.0);
+  // 9007199254743568 can be represented exactly, but will be clamped to
+  // 9007199254743560
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254743568.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::UnconditionalAKAHighRes);
+  ASSERT_EQ(result, 9007199254743560.0);
+  cleanupJitter(jitterEnabled);
+}
+
+TEST(ResistFingerprinting, ReducePrecision_ExpectedLossOfPrecision_RFP)
+{
+  bool jitterEnabled = setupJitter(false);
+  double result;
+  // We lose integer precision at 9007199254740992 - let's confirm that.
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740992.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::RFP);
+  ASSERT_EQ(result, 9007199254740990.0);
+  // 9007199254740995 is approximated to 9007199254740980
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740995.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::RFP);
+  ASSERT_EQ(result, 9007199254740995);
+  // 9007199254740999 is approximated as 9007199254741000
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740999.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::RFP);
+  ASSERT_EQ(result, 9007199254741000.0);
+  // 9007199254743568 can be represented exactly, but will be clamped to
+  // 9007199254743560
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254743568.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::RFP);
+  ASSERT_EQ(result, 9007199254743565.0);
+  cleanupJitter(jitterEnabled);
+}
+
+TEST(ResistFingerprinting,
+     ReducePrecision_ExpectedLossOfPrecision_DangerouslyNone)
+{
+  bool jitterEnabled = setupJitter(false);
+  double result;
+  // We lose integer precision at 9007199254740992 - let's confirm that.
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740992.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 9007199254740992.0);
+  // 9007199254740995 is approximated to 9007199254740980
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740995.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 9007199254740995);
+  // 9007199254740999 is approximated as 9007199254741000
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254740999.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 9007199254740999.0);
+  // 9007199254743568 can be represented exactly, but will be clamped to
+  // 9007199254743560
+  result = nsRFPService::ReduceTimePrecisionImpl(
+      9007199254743568.0, nsRFPService::TimeScale::MicroSeconds, 5, -1,
+      TimerPrecisionType::DangerouslyNone);
+  ASSERT_EQ(result, 9007199254743568.0);
+  cleanupJitter(jitterEnabled);
+}
+
 // Use an ugly but simple hack to turn an integer-based rand()
 // function to a double-based one.
 #define RAND_DOUBLE (rand() * (rand() / (double)rand()))
@@ -296,98 +474,98 @@
 
   result = nsRFPService::ReduceTimePrecisionImpl(
       1, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 0);
   result = nsRFPService::ReduceTimePrecisionImpl(
       327, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 0);
   result = nsRFPService::ReduceTimePrecisionImpl(
       328, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       329, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       499, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 500);
 
   result = nsRFPService::ReduceTimePrecisionImpl(
       500, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       540, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       547, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       548, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 1000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       930, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 1000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       1255, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 1000);
 
   result = nsRFPService::ReduceTimePrecisionImpl(
       4000, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4220, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4224, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4225, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4340, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4499, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4500);
 
   result = nsRFPService::ReduceTimePrecisionImpl(
       4500, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4536, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4695, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 4500);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4698, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 5000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       4726, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 5000);
   result = nsRFPService::ReduceTimePrecisionImpl(
       5106, nsRFPService::TimeScale::MicroSeconds, 500, 4000,
-      TimerPrecisionType::All);
+      TimerPrecisionType::Normal);
   ASSERT_EQ(result, 5000);
 
   cleanupJitter(jitterEnabled);
